World renderer lwjgl refactoring

emanuele3d opened the pull request:
* Removed Physics-related code from WorldRenderer. 
* Methods are now using explicitly the activeCamera or the shadowMapCamera instance members rather than a camera passed in as input, if appropriate (1). Some methods acts on both and have been left untouched. Also both member variables are now final, for the time being.
* Moved AABB visibility testing from WorldRenderer to Camera class. 
* WorldRendererLwjgl class no longer providing the number of mesh segments per vertical chunk: the value can be found through the config, other affected classes have been modified accordingly. 
* WorldRendererLwjgl now has explicit variables initialized early for the three configs it uses. 
* Removed unnecessary comments among head-of-file variable declarations. 
* Apart from some other small changes, all other changes are consequences.

(1) This could be considered a step back given that I agree eventually Cameras should be components handled by the ES. I just felt that for the current state of the code this way is a bit clearer.

This is just a warm up PR, the next ones will be more conceptually substantial.

emanuele3d commented on engine/src/main/java/org/terasology/engine/modes/loadProcesses/InitialiseWorld.java:
```diff
@@ -147,8 +147,6 @@ public boolean step() {
         CoreRegistry.put(LocalPlayer.class, new LocalPlayer());
         CoreRegistry.put(Camera.class, worldRenderer.getActiveCamera());
-        CoreRegistry.put(PhysicsEngine.class, worldRenderer.getBulletRenderer());
-        CoreRegistry.put(Physics.class, worldRenderer.getBulletRenderer());
```
Physics initialization is already handled in LoadProcess InitializePhysics.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/cameras/Camera.java:
```diff
@@ -237,4 +238,8 @@ public void setzFar(float zFar) {
         return reflected;
     }
+
+    public boolean hasInSight(AABB aabb) {
```
Because of its frustrum-dependency, this visibility testing I felt was appropriate moved into the Camera class rather than the WorldRenderer.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/world/ChunkMeshUpdateManager.java:
```diff
@@ -133,6 +135,8 @@ public void shutdown() {
 
     private static class ChunkUpdateTask implements ChunkTask {
 
+        private final int numberOfMeshSegments = CoreRegistry.get(Config.class).getSystem().getVerticalChunkMeshSegments();
```
Due to previous discussion with Immortius, I couldn't set this as a static variable, as we cannot be sure if the CoreRegistry will have the right entry by the time this class is touched.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/world/WorldRendererLwjgl.java:
```diff
@@ -193,31 +162,26 @@ public boolean pregenerateChunks() {
-    public void update(float delta) {
+    public void update(float deltaInMs) {
 
         PerformanceMonitor.startActivity("Update Tick");
```
I have been puzzled for some time about these PerformanceMonitor statements. This one in particular is a bit strange as I doubt the updateTick() method is of any relevance from a Performance perspective, as it is currently implemented (tick += deltaInMs * 1000;). 

Specifically, I intend as much as possible to place them -within- methods rather than around the method call. But the overarching question is, what should be monitored this way and what shouldn't? And isn't this monitoring the role of a standard profiler anyway rather than the code itself?

immortius replied:
float deltas are in seconds, not milliseconds. So the renamed parameter is incorrect.

I threw performance monitoring around most methods in the renderer to determine where time was spent. Yes, updateTick() probably doesn't have much cost, but I wasn't going to assume that. Feel free to strip some out as you go. In general I guess the approach is to have the monitor around some top level method, and only drilling down if more information is needed.

I put them around methods rather than in methods simply because methods can have multiple return locations - having them around the methods avoid the need for try/finally or multiple endActivity statements, and basically dodges maintenance errors that introduce new return locations.

The Performance monitor is different from profiling. Profiling will tell you how much time is spent in a given method over the life of the application, but we're more interested in how time is spent each frame, and if a frame is unusually slow - why was it slow? If a method takes a second once in the running of an application then that won't really show in the profile statistics - but that would be very noticeable for a user. It also drives the in-game monitoring metrics, which means a user can report where time is spent for them without them having to set up to run from source.

emanuele3d replied:
I'll rename the parameter appropriately. This prompts a question though: why does the animation system needs times in 1000s of seconds?

Thank you for the information about PerformanceMonitor, now I understand its purpose and usage much better. For what I have seen in the code there isn't a huge amount of multiple return location. But I didn't think about this aspect before and I'll now double-check to make sure that each startActivity is appropriately matched by an endActivity. Given that this is the only concern on the matter however, I will generally move those statements within methods. I really feel that they fragment the code too much, especially the high-level methods. And the issue of multiple return location is probably easily addressable within this area of the codebase.

emanuele3d replied:
Another long-shot thought on the issue of multiple return locations:  python has a construct in the form:

    with(lock) {
   }

which allows to automatically release a lock whenever the code within the block returns or generates an exception. Is there any such thing in Java and could the PerformanceMonitor be made to behave similarly?

immortius replied:
delta is in seconds (e.g. 0.02 seconds for a frame at 50 frames a second).
delta * 1000 is in milliseconds (e.g. 0.02 * 1000 = 20 milliseconds).
So tick is in milliseconds. Not sure how it is used, but I guess it is a remainder from my original sweep of changing all deltas from milliseconds to seconds (easier to work with for game logic IMO).

You can use ```try (Closable c = new Closable()) {}``` blocks in Java. The problem is the PerformanceMonitor framework was created before we switched to Java 7 that introduced it, so hasn't been set up so that activities are a Closable object. I could add that though.

The longer winded alternative is:
```java
PerformanceMonitor.startActivity("Fun");
try {
  // ....
} finally {
  PerformanceMonitor.endActivity();
}```

emanuele3d replied:
I double-checked the instances of PerformanceMonitor usage in WorldRendererLwjgl and RenderableWorld. I have not found any that framed a section of code with multiple return location, so no need for special handling so far.

The use of Closable objects to wrap sections of code is interesting though. I remember changing the GameEngine interface to extend AutoCloseable, which I seem to understand is similar. I wonder how difficult it would be to change the PerformanceMonitor to work with it. I'd rather not do it for the time being though. I have enough on my plate already.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/world/WorldRendererLwjgl.java:
```diff
@@ -442,14 +406,14 @@ public void renderWorld(Camera camera) {
         DefaultRenderingProcess.getInstance().beginRenderLightGeometryStencilPass();
         Material program = Assets.getMaterial("engine:prog.simple");
```
I can understand where an asset "engine:simple" is located. What I didn't manage to understand yet, even looking at the Asset Management code, is where is that "prog." substring is handled and why is it necessary.

immortius replied:
There is actually no "engine:simple" material asset. There is an "engine:simple" shader. The lwjgl shader manager generates materials for a number of shaders used as part of the rendering, and to ensure these materials don't conflict with any in the assets folders their names are prefixed with "prog."

emanuele3d replied:
So, the Asset Manager generates at runtime a Material "engine:prog.simple" out of the shader "engine:simple". And the way to reference the material associated with a shader (rather than the shader itself) is to insert the "prog." string within the asset URI. Correct? 

Wouldn't this more appropriately handled via some call on the shader? I.e. shader.getMaterial() which in turn might generate one if one isn't available yet?

immortius replied:
Not the asset manager. ShaderManagerLwjgl. These are not automatically generated assets based on a naming convention! These are specific materials generated by the rendering system. They are also purely for the internal usage of the rendering system - these are the materials used by the ShaderParameter classes. Ideally we would eventually phase them out, replaced with a shader variable semantics support or something, but because that is quite a body of work not really worth messing with.

There is not a one to one relationship between materials and shaders, and in general a shader does not have enough information to generate a material (since it doesn't have values for the shader's variables - this is why a material exists). A shader is included in the MaterialData class used to generate a material. So if you want a material for a given shader, you make it.

emanuele3d replied:
Thank you for your patience Immortius and explaining me this. I'll look into the ShaderManagerLwjgl and Material classes to understand the details. Indeed I do not forsee the need to change anything in this department at this stage. But I probably need to understand better how it works.

Cervator commented:
Tests out fine, not that I have much of anything specific to test. But the game works! :-)

For the monitoring it is nice to have it in the client since you can use F3 + cycle F4 to see stats on any client. Whether the start/stop calls are inside or outside methods I've got no opinion on, wherever seems sensible.

Rest is above my head, but looks like progress :-)

@immortius can probably answer the "prog" thing and give any other feedback. But you're probably making steady progress in an area you're quickly becoming the leading expert on anyway!

immortius commented on engine/src/main/java/org/terasology/rendering/world/WorldRendererLwjgl.java:
```diff
@@ -78,57 +76,37 @@
 
-    /* CAMERA */
-    private Camera activeCamera;
+    private final Camera activeCamera;
```
This seems a little odd - activeCamera implies the camera can change, but final will prevent this?

emanuele3d replied:
In short: you are right. 

Right now I don't think the code can handle switching cameras. For one there is no setActiveCamera() method yet: the camera is created in the constructor and stay the same. So, I just formalized what I think it's the state of affair. Also, RenderableWorld has a pregenerateChunk() method that is only called on world loading time. The method RenderableWorldImpl.updateChunkInProximity() might be enough to regenerate the world around a new camera, but I need to verify that. For sure, RenderableWorld currently has no facility to switch cameras.

That been said, it is something I'd like to sort out. I just need to understand better what needs to be updated when the camera changes. For the time being I can:

1. rename the camera to something like "playerCamera"
2. remove the final keyword
3. add a TODO and leave it as it is, to be worked on in one of the next PRs

What do you prefer?

immortius replied:
Since we don't need camera switching presently, I'ld just go with (1).

emanuele3d replied:
Done, in the next commit.

emanuele3d replied:
Added note: WorldRenderer has a getActiveCamera() method. This is used in many, many other files and for the time being I'd rather leave it as it is rather than rename it getPlayerCamera(). Hopefully in one of the next PRs I'll be able to rename the camera in the code back to activeCamera anyway, by making the code capable of switching cameras at runtime.

emanuele3d commented:
* Renamed appropriately delta parameter in WorldRendererLwjgl.update(float delta)
* Renamed activeCamera to playerCamera in both WorldRendererLwjgl and renderableWorldImpl classes. Interfaces (getActiveCamera()) left untouched for the time being. 
* Double-checked all usages of PerformanceMonitor in the aformentioned classes for unhandled multiple return locations: none were found.

Cervator commented:
Is this ready for merging then? :-)

emanuele3d commented:
From my perspective it is. The only thing that might be in question is renaming that getActiveCamera() method to getPlayerCamera() method, buy I'd advise against as it'd be only temporary. @immortius?

emanuele3d commented:
Shut, I just noticed the merge conflicts that need resolving...

emanuele3d commented:
* merged latest develop in the branch
* addressed conflicts in WorldRendererLwjgl
* taken onboard @msteiger suggestion regarding transformPoint vs transformVector.

Cervator commented:
Merged! 

Might be time to try your luck at rebase instead of merge soon :D

emanuele3d commented:
Did you say the R word? =D

Thanks for reminding me. I'll go through the docs about Rebasing again and
I'll try next time!

On 8 January 2015 at 06:18, Rasmus Praestholm <notifications@github.com>
wrote:

> Merged!
>
> Might be time to try your luck at rebase instead of merge soon :D
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/MovingBlocks/Terasology/pull/1481#issuecomment-69137221>
> .
>
